packages/preferences/src/browser/preference-tree-model.ts

```diff
@@ -0,0 +1,185 @@
+/********************************************************************************
+ * Copyright (C) 2020 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, postConstruct } from 'inversify';
+import {
+    TreeModelImpl,
+    TreeWidget,
+    CompositeTreeNode,
+    TopDownTreeIterator,
+    TreeNode,
+    PreferenceSchemaProvider,
+    PreferenceDataProperty,
+    NodeProps,
+    ExpandableTreeNode
+} from '@theia/core/lib/browser';
+import { Emitter } from '@theia/core';
+import { PreferencesSearchbarWidget } from './views/preference-searchbar-widget';
+import { PreferenceTreeGenerator } from './util/preference-tree-generator';
+import * as fuzzy from 'fuzzy';
+import { PreferencesScopeTabBar } from './views/preference-scope-tabbar-widget';
+import { Preference } from './util/preference-types';
+import { Event } from '@theia/core/src/common';
+
+export interface PreferenceTreeNodeRow extends TreeWidget.NodeRow {
+    visibleChildren: number;
+    isExpansible?: boolean;
+}
+export interface PreferenceTreeNodeProps extends NodeProps {
+    visibleChildren: number;
+    isExpansible?: boolean;
+}
+
+@injectable()
+export class PreferenceTreeModel extends TreeModelImpl {
+
+    @inject(PreferenceSchemaProvider) protected readonly schemaProvider: PreferenceSchemaProvider;
+    @inject(PreferencesSearchbarWidget) protected readonly filterInput: PreferencesSearchbarWidget;
+    @inject(PreferenceTreeGenerator) protected readonly treeGenerator: PreferenceTreeGenerator;
+    @inject(PreferencesScopeTabBar) protected readonly scopeTracker: PreferencesScopeTabBar;
+
+    protected readonly onTreeFilterChangedEmitter = new Emitter<{ filterCleared: boolean; rows: Map<string, PreferenceTreeNodeRow>; }>();
+    readonly onFilterChanged = this.onTreeFilterChangedEmitter.event;
+
+    protected lastSearchedFuzzy: string = '';
+    protected lastSearchedLiteral: string = '';
+    protected _currentScope: number = Number(Preference.DEFAULT_SCOPE.scope);
+    protected _isFiltered: boolean = false;
+    protected _currentRows: Map<string, PreferenceTreeNodeRow> = new Map();
```

Comment:
The tree model should not had such representation. Why do you need it here? Flat rows should be computed in the widget based on nodes' attributes like `visible`.

Reply:
Because this tree - and the rules about what should and shouldn't be displayed - are common to the `PreferenceTreeWidget` and `PreferenceEditorWidget` it's more convenient and more efficient to simply share that data in a location accessible to both widgets. Similarly, scroll actions in the `PereferenceEditorWidget` affect the expansion and selection states of the `PreferenceTreeWidget`, so sharing the model between the two of them allows the `PreferenceEditorWidget` to implement that logic while the `PreferenceTreeWidget` is able to simply react to the model's selection and expansion events normally.